src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/transformers/SnapshotTrackerTransformer.kt

```diff
@@ -0,0 +1,235 @@
+
+            if (matchedArguments.isEmpty()) {
+                visitMethodInsn(opcode, owner, name, desc, itf)
+                return
+            }
+
+            invokeIfInTestingCode(
+                original = { visitMethodInsn(opcode, owner, name, desc, itf) },
+                code = {
+                    // STACK: args
+                    val arguments = storeArguments(desc)
+                    val matchedLocals = arguments.filterIndexed { index, _ -> matchedArguments.contains(index) }
+                    // STACK: <empty>
+                    push(matchedLocals.size)
+                    // STACK: length
+                    visitTypeInsn(ANEWARRAY, "java/lang/Object")
+                    // STACK: array
+                    matchedLocals.forEachIndexed { index, local ->
+                        // STACK: array
+                        visitInsn(DUP)
+                        push(index)
+                        loadLocal(local)
+                        // STACK: array, array, index, obj
+                        visitInsn(AASTORE)
+                        // STACK: array
+                    }
+                    // STACK: array
+                    invokeStatic(Injections::updateSnapshotWithEnergeticTracking)
+                    // STACK: <empty>
+                    loadLocals(arguments)
+                    // STACK: args
+                    visitMethodInsn(opcode, owner, name, desc, itf)
+                }
+            )
+        }
+        else visitMethodInsn(opcode, owner, name, desc, itf)
+    }
+
+    override fun visitFieldInsn(opcode: Int, owner: String, fieldName: String, desc: String) = adapter.run {
+        if (
+            isCoroutineInternalClass(owner) ||
+            isCoroutineStateMachineClass(owner) ||
+            // when initializing our own fields in constructor, we do not want to track that as snapshot modification
+            (methodName == "<init>" && className == owner)
+        ) {
+            visitFieldInsn(opcode, owner, fieldName, desc)
+            return
+        }
+
+        when (opcode) {
+            GETSTATIC, PUTSTATIC -> {
+                // STACK: [<empty> | value]
+                invokeIfInTestingCode(
+                    original = {
+                        visitFieldInsn(opcode, owner, fieldName, desc)
+                    },
+                    code = {
+                        // STACK: [<empty> | value]
+                        pushNull()
+                        push(owner.canonicalClassName)
+                        push(fieldName)
+                        loadNewCodeLocationId()
+                        // STACK: [<empty> | value], null, className, fieldName, codeLocation
+                        invokeStatic(Injections::updateSnapshotOnFieldAccess)
+                        // STACK: [<empty> | value]
+                        visitFieldInsn(opcode, owner, fieldName, desc)
+                        // STACK: [<empty> | value]
+                    }
```

Comment:
This code is very similar to the code in `SharedMemoryAcessTransformer`. Let's unify the two and use only `SharedMemoryAccessTransformer`.

Reply:
done